### PR TITLE
hardened sendKeys()

### PR DESCRIPTION
**Pull request by pbruski**

### CODE REVIEW

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/ElementUtils.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package com.atlassian.performance.tools.jiraactions.api.webdriver
+
+import com.atlassian.performance.tools.jiraactions.api.page.wait
+import org.openqa.selenium.By
+import org.openqa.selenium.Keys
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.WebElement
+import org.openqa.selenium.support.ui.ExpectedConditions
+
+class ElementUtils {
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/ElementUtils.kt]:
```suggestion
class HardenedKeyboard {
```

- [ ] consider a less vague name

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/LoginPage.kt [hunk 2]
```diff
@@ -19,9 +20,10 @@ class LoginPage(
             ExpectedConditions.presenceOfElementLocated(loginFormLocator)
         )
         val loginForm = driver.findElement(loginFormLocator)
-        loginForm.findElement(By.name("os_username")).sendKeys(user.name)
-        loginForm.findElement(By.name("os_password")).sendKeys(user.password)
+
+        ElementUtils.sendKeysAndValidate(driver, By.name("os_username"), user.name)
+        ElementUtils.sendKeysAndValidate(driver, By.name("os_password"), user.password)
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/LoginPage.kt]: It's a perfect place for an [extension function](https://kotlinlang.org/docs/reference/extensions.html).
It'd read:
```suggestion
        loginForm.findElement(By.name("os_password")).sendKeysAndValidate(user.password)
```

Extension functions compile to statics:
image

Upsides:
* easier on the eyes
* applicable to any `WebElement`, not only the ones from `WebDriver.findElement`

- [ ] consider extension functions

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/ElementUtils.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+            element.sendKeys(*keysToSend)
+            return element
+        }
+
+        @JvmStatic
+        fun sendKeysAndValidate(driver: WebDriver, locator: By, text: String): WebElement {
+            val element = sendKeysWhenClickable(driver, locator, text as CharSequence)
+
+            // It's hard to say when the keys can be sent. They seem to randomly get lost.
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/ElementUtils.kt]: Truth be told, sometimes I encounter https://ecosystem.atlassian.net/browse/JPERF-298 as a real user.
When I tab and type too quickly, beginning of my inputs get dropped.
Such are our frontends right now.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/page/form/TextSelectField.kt [hunk 4]
```diff
@@ -24,8 +24,7 @@ internal class TextSelectField(
 
     override fun fillWithAnyValue() {
         val component = random.pick(select.options)!!.getAttribute("innerHTML").trim()
-        input.sendKeys(component)
-        input.sendKeys(Keys.RETURN)
+        input.sendKeys(component, Keys.RETURN)
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/page/form/TextSelectField.kt]: Nice

File: src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/scenario/JiraCoreScenarioIT.kt [hunk 5]
```diff
@@ -144,19 +143,4 @@ class JiraCoreScenarioIT {
             throw e
         }
     }
-
-    private fun sendKeys(driver: RemoteWebDriver, locator: By, text: String) {
```
Comment (dagguh) [inline src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/scenario/JiraCoreScenarioIT.kt]: It's always nice to see low-level stuff get separated from high-level stuff 👍

File: CHANGELOG.md [hunk 6]
```diff
@@ -26,6 +26,9 @@ Dropping a requirement of a major version of a dependency is a new contract.
 ## [Unreleased]
 [Unreleased]: https://github.com/atlassian/jira-actions/compare/release-3.10.0...master
 
+### Added
+- Hardened sendKeys() method: ElementUtils#sendKeysAndValidate
```
Comment (wyrzyk) [inline CHANGELOG.md]: No public issue key? 😢

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/LoginPage.kt (hunk 2 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/LoginPage.kt]: It's a little bit easier to read and use in kotlin. Will the extension function be also accessible for java consumers? (as it's a part of API)

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/ElementUtils.kt (hunk 1 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/ElementUtils.kt]: If we'll harden it more it'll become `EvenMoreHardenedKeyboard` :)

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/ElementUtils.kt (hunk 3 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/ElementUtils.kt]: 👍

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/LoginPage.kt (hunk 2 above)
Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/LoginPage.kt]: Yes, they'd be available as static methods.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/LoginPage.kt [hunk 7]
```diff
@@ -19,9 +20,11 @@ class LoginPage(
             ExpectedConditions.presenceOfElementLocated(loginFormLocator)
         )
         val loginForm = driver.findElement(loginFormLocator)
-        loginForm.findElement(By.name("os_username")).sendKeys(user.name)
-        loginForm.findElement(By.name("os_password")).sendKeys(user.password)
+
+        driver.findElement(By.name("os_username")).sendKeysAndValidate(driver, user.name)
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/LoginPage.kt]: Is there a reason for switching from `loginForm.findElement` to `driver.findElement`?

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/WebElementExt.kt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        if (actualValue == text) {
+            break
+        }
+        Thread.sleep(100)
+        sendKeys(Keys.BACK_SPACE.repeat(actualValue.length), text)
+    }
+    return this
+}
+
+fun WebElement.getHtmlValue(): String? = getAttribute("value")
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/WebElementExt.kt]: In general it's not worth to introduce such shallow wrappers.
Next we'd add `getHref`, `getId`, `getName`, `getType`, `getLabel`, etc.

`getHtmlValue` poses more questions than it answers - what is an HTML value? Is it a `form` value? Is it `innerHTML`?
The answer is: it's an attribute. Using `getAttribute("value")` directly would be clear from the get-go.

- [ ] consider inlining super-thin wrappers

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/WebElementExt.kt [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package com.atlassian.performance.tools.jiraactions.api.webdriver
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/webdriver/WebElementExt.kt]: `WebElementExt.kt` -> `HardenedKeyboard.kt`

This way it won't attract random behavior.

**Pull request merged**